core/src/com/unciv/logic/city/CityConstructions.kt

```diff
@@ -319,7 +323,26 @@ class CityConstructions {
                     || rejectionReason.startsWith("Cannot be built with")
                     || rejectionReason.startsWith("Don't need to build any more")
                     || rejectionReason.startsWith("Obsolete")
-            ) inProgressConstructions.remove(constructionName)
+            ) {
+                if (construction is Building) {
+                    // Production put into wonders gets refunded
+                    if (construction.isWonder && getWorkDone(constructionName) != 0) {
+                        cityInfo.civInfo.gold += getWorkDone(constructionName)
+                        val buildingIcon = "BuildingIcons/${constructionName}"
+                        cityInfo.civInfo.addNotification("Could not continue work on [${constructionName}], excess production converted to [${getWorkDone(constructionName)}] gold", NotificationIcon.Gold, buildingIcon)
+                    }
+                } else if (construction is BaseUnit) {
+                    // Production put into upgradable units gets put into upgraded version
+                    if (rejectionReason.startsWith("Obsolete") && construction.upgradesTo != null) {
+                        if (!inProgressConstructions.contains(construction.upgradesTo)) {
+                            inProgressConstructions[construction.upgradesTo!!] = getWorkDone(constructionName)
+                        } else {
+                            inProgressConstructions[construction.upgradesTo!!] = inProgressConstructions[construction.upgradesTo!!]!! + getWorkDone(constructionName)
```

Comment:
Kotlin does have `+=` (feel free to ignore if that's just your style)

Reply:
Previously I used the `+=` but got an error because `inProgressConstructions[construction.upgradesTo!!]` can return `null` and in that case the `+=` operator is not defined. What I have now is my workaround for that. Is there a good/standard way to work around this?

Reply:
(why does it say 'reviewed'? Did I hit the wrong button?)
> `+=`

Right. My bad. :crying_cat_face:   ...Then the Elvis<sup>®</sup> is the most 'Kotlin'-elegant short form... or `map[key] = delta + (if (key in map) map[key]!! else 0)` or... All versions including yours probably give the same bytecode or close.

Tempts one to overload HashMap to allow adding to nonexisting entries directly - oh, wait, that's [already](https://github.com/yairm210/Unciv/blob/master/core/src/com/unciv/models/Counter.kt) been done in Unciv... But changing the `inProgressConstructions` type could break serialization.